### PR TITLE
145512- implement additional coverage scenario

### PR DESCRIPTION
**Pull request by sakshi15525**
1. Added action configuration for additional coverage scenario.
2. Added json to xml mapping.
3. Rearranged the sequence of all actions.
4. Added unit test cases.
5. Added automation scripting.

### CODE REVIEW

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 1]
```diff
@@ -613,7 +665,7 @@ private static bool VerifyCreateENCCell(string childCell, ZMAT_ACTIONITEMS item,
 
         private static bool VerifyDecryptedPermit(string childCell, ZMAT_ACTIONITEMS item, string permitState)
         {
-            Console.WriteLine("Action#:" + ActionCounter + ".Childcell:" + childCell);
+            //Console.WriteLine("Action#:" + ActionCounter + ".Childcell:" + childCell);
```
Comment (pravinvarade10801) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:668]: If commented line is not required so can we remove this?

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 2]
```diff
@@ -613,7 +665,7 @@ private static bool VerifyCreateENCCell(string childCell, ZMAT_ACTIONITEMS item,
 
         private static bool VerifyDecryptedPermit(string childCell, ZMAT_ACTIONITEMS item, string permitState)
```
Comment (pravinvarade10801) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:666]: As console line is commented in this method so childCell parameter looks not used any more, can we remove this?

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 3]
```diff
@@ -992,6 +1045,21 @@ public static int CalculateReplaceCellActionCount(JsonPayloadHelper jsonPayload)
 
             return count;
         }
+        public static int CalculateAdditionalcoverageCellActionCount(JsonPayloadHelper jsonPayload)
```
Comment (pravinvarade10801) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:1048]:
```suggestion
        public static int CalculateAdditionalCoverageCellActionCount(JsonPayloadHelper jsonPayload)
```

**Review by pravinvarade10801 — COMMENTED** [review]
Have added few minor comments please.

**Review by Sadha1501493 — CHANGES_REQUESTED** [review]

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 4]
```diff
@@ -450,6 +456,52 @@ public static async Task<bool> CheckXMLAttributes(JsonPayloadHelper jsonPayload,
...
+                    if (!item.PRODTYPE.Equals(product.ProductType[4..]))
+                        AttrNotMatched.Add(nameof(item.PRODTYPE));
+                    if (!item.WEEKNO.Equals(weekNoTag))
+                        AttrNotMatched.Add(nameof(item.WEEKNO));
+                    if (!item.VALIDFROM.Equals(validFromTag))
+                        AttrNotMatched.Add(nameof(item.VALIDFROM));
+                    if (!item.CORRECTION.Equals(correctionTag))
+                        AttrNotMatched.Add(nameof(item.CORRECTION));
+                    //Checking blanks
+                    List<string> blankFieldNames = new List<string> { "PRODUCTNAME","CANCELLED", "AGENCY", "PROVIDER", "ENCSIZE", "TITLE", "EDITIONNO", "UPDATENO", "UNITTYPE", "ACTIVEKEY", "NEXTKEY" };
```
[the remark this answers is not in the document]
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs]: Done

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs (hunk 2 above)
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs]: done

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs (hunk 3 above)
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs]: Done

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID38_1AdditionalCoverageCell.JSON [hunk 5]
```diff
@@ -0,0 +1,182 @@
+{
```
[the remark this answers is not in the document]
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID38_1AdditionalCoverageCell.JSON]: done

File: tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs [hunk 6]
```diff
@@ -112,7 +112,10 @@ public async Task WhenValidEventInNewEncContentPublishedEventReceivedWithTokenHa
 
         // Rule change for create avcs unit of sale (having multiple products in addProducts)
         [TestCase("ID37_CreateUoSHavingMultipleItemsInAddProducts.JSON", "N", TestName = "WhenICallTheWebhookWithMoveAndNewCellScenarioWhereUoSHasMultipleValuesInAddProducts_ThenWebhookReturns200Response")]
-
+        [TestCase("ID38_1AdditionalCoverageCell.JSON", "N", TestName = "WhenICallTheWebhookWithOneAdditionalCoverageCellScenario_ThenWebhookReturns200Response")]
```
[the remark this answers is not in the document]
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs]: done

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID39_1AdditionalCoverageWithReplacedCell.JSON [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+                "isNewUnitOfSale": false,
+                "_COMMENT": "avcs_cat has either boundingBox or polygon - should both be included?",
+                "geographicLimit": {
+                    "boundingBox": {
+                        "northLimit": 22.643255,
+                        "southLimit": 22.4625767,
+                        "eastLimit": 120.34972,
+                        "westLimit": 120.219475
+                    },
+                    "polygons": [
```
Comment (Sadha1501493) [inline tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID39_1AdditionalCoverageWithReplacedCell.JSON:265]: Please remove Polygon section

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+                "replacedBy": [],
+                "additionalCoverage": [],
+                "geographicLimit": {
+                    "boundingBox": {
+                        "northLimit": 63,
+                        "southLimit": 63,
+                        "eastLimit": 162,
+                        "westLimit": 161
+                    },
+                    "polygons": [
```
Comment (Sadha1501493) [inline tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON:178]: Please remove Polygon section

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+                ],
+                "_COMMENT": "BoundingBox or polygon or both or neither",
+                "geographicLimit": {
+                    "boundingBox": {
+                        "northLimit": 24.146815,
+                        "southLimit": 22.581615,
+                        "eastLimit": 120.349635,
+                        "westLimit": 119.39142
+                    },
+                    "polygons": []
```
Comment (Sadha1501493) [inline tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON:253]: Please remove Polygon section

**Review by pravinvarade10801 — APPROVED** [review]
LGTM

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+                "isNewUnitOfSale": false,
+                "_COMMENT": "avcs_cat has either boundingBox or polygon - should both be included?",
+                "geographicLimit": {
+                    "boundingBox": {
+                        "northLimit": 22.643255,
+                        "southLimit": 22.4625767,
+                        "eastLimit": 120.34972,
+                        "westLimit": 120.219475
+                    },
+                    "polygons": [
```
Comment (Sadha1501493) [inline tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON:280]: Please remove Polygon section

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+                    "unit",
+                    "folio"
+                ],
+                "unitSize": "large",
+                "unitType": "AVCS Online Folio",
+                "status": "ForSale",
+                "isNewUnitOfSale": false,
+                "geographicLimit": {
+                    "boundingBox": {},
+                    "polygons": [
```
Comment (Sadha1501493) [inline tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON:305]: Please remove Polygon section

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+                "unitOfSaleType": "folio",
+                "_ENUM": [
+                    "unit",
+                    "folio"
+                ],
+                "status": "ForSale",
+                "isNewUnitOfSale": false,
+                "geographicLimit": {
+                    "boundingBox": {},
+                    "polygons": [
```
Comment (Sadha1501493) [inline tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON:350]: Please remove Polygon section

**Review by Sadha1501493 — CHANGES_REQUESTED** [review]

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 13]
```diff
@@ -428,6 +432,47 @@ public static async Task<bool> CheckXMLAttributes(JsonPayloadHelper jsonPayload,
...
+                    VerifyBlankFields(item, blankFieldNames);
+
+                    if (AttrNotMatched.Count == 0)
+                    {
+                        Console.WriteLine("ADDITIONAL COVERAGE ENC CELL Action's Data is correct");
+                        int valueIndex = product.AdditionalCoverage.IndexOf(replaceBy);
+                        product.AdditionalCoverage[valueIndex] = product.AdditionalCoverage[valueIndex].Replace(replaceBy, "skip");
+                        return true;
+                    }
+                    else
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:462]: else keyword not required.

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 14]
```diff
@@ -428,6 +432,47 @@ public static async Task<bool> CheckXMLAttributes(JsonPayloadHelper jsonPayload,
...
+                    if (!item.ACTIONNUMBER.Equals(ActionCounter.ToString()))
+                        AttrNotMatched.Add(nameof(item.ACTIONNUMBER));
+                    //xmlAttributes[1] is skipped as already checked
+                    if (!item.PRODUCT.Equals("ENC CELL"))
+                        AttrNotMatched.Add(nameof(item.PRODUCT));
+                    if (!item.PRODTYPE.Equals(product.ProductType[4..]))
+                        AttrNotMatched.Add(nameof(item.PRODTYPE));
+                    VerifyAdditionalXmlTags(item, correctionTag);
+                    //Checking blanks
+                    List<string> blankFieldNames = new List<string> {"PRODUCTNAME", "CANCELLED", "AGENCY", "PROVIDER", "ENCSIZE", "TITLE", "EDITIONNO", "UPDATENO", "UNITTYPE", "ACTIVEKEY", "NEXTKEY" };
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:452]:
```suggestion
                    List<string> blankFieldNames = new() {"PRODUCTNAME", "CANCELLED", "AGENCY", "PROVIDER", "ENCSIZE", "TITLE", "EDITIONNO", "UPDATENO", "UNITTYPE", "ACTIVEKEY", "NEXTKEY" };
```

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 15]
```diff
@@ -964,6 +1007,21 @@ public static int CalculateReplaceCellActionCount(JsonPayloadHelper jsonPayload)
 
             return count;
         }
+        public static int CalculateAdditionalCoverageCellActionCount(JsonPayloadHelper jsonPayload)
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:1010]: Add blank line before this method

**Review by Vishal14583 — CHANGES_REQUESTED** [review]
Few minor review comments added.

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs (hunk 15 above)
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:969]: Done

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs (hunk 14 above)
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs]: done

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs (hunk 13 above)
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs]: done

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID39_1AdditionalCoverageWithReplacedCell.JSON [hunk 16]
```diff
@@ -0,0 +1,302 @@
+{
+    "specversion": "1.0",
+    "type": "uk.gov.ukho.encpublishing.enccontentpublished.v2.1",
+    "source": "https://encpublishing.ukho.gov.uk",
+    "id": "2f03a25f-28b3-46ea-b009-5943250a9a41",
+    "time": "2020-10-13T12:08:03.4880776Z",
+    "_COMMENT": "A comma separated list of products",
```
[the remark this answers is not in the document]
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID39_1AdditionalCoverageWithReplacedCell.JSON]: done

**Review by Vishal14583 — APPROVED** [review]
LGTM

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON (hunk 12 above)
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON]: Done

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON (hunk 11 above)
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON]: done

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON (hunk 10 above)
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON]: done

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON (hunk 8 above)
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID40_1AdditionalCoverageWithReplacedCellAnd1CancelledCell.JSON]: done

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID39_1AdditionalCoverageWithReplacedCell.JSON (hunk 7 above)
Comment (MitaliSingh16) [reply tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/ID39_1AdditionalCoverageWithReplacedCell.JSON]: done

**Review by jrippington — APPROVED** [review]
Looks good to me

**Pull request merged**